src/almo_scf_optimizer.F

```diff
@@ -2704,91 +2708,120 @@ SUBROUTINE almo_scf_construct_nlmos(qs_env, optimizer, &
-                        END SELECT
-                        localization_obj_function_ispin = localization_obj_function_ispin + fval
-                     ENDDO
-
-                  ENDDO ! end loop over idim0
-
-                  DEALLOCATE(z2)
-                  DEALLOCATE(reim_diag)
-
-                  localization_obj_function_ispin = localiz_coeff * localization_obj_function_ispin
-                  localization_obj_function = localization_obj_function + localization_obj_function_ispin
-
-               ENDIF
-
-               IF (penalty_occ_vol) THEN
-
-                  CALL dbcsr_multiply("N", "N", 1.0_dp, &
-                          m_S0(ispin), &
-                          m_theta_normalized(ispin), &
-                          0.0_dp, tempOccOcc1(ispin), &
-                          filter_eps=eps_filter)
-                  ! compute current sigma
-                  CALL dbcsr_multiply("T", "N", 1.0_dp, &
-                          m_theta_normalized(ispin), &
-                          tempOccOcc1(ispin), &
-                          0.0_dp, m_sigma(ispin), &
-                          filter_eps=eps_filter)
-
-                  CALL determinant(m_sigma(ispin), det1, &
-                          eps_filter)
-                  ! save the current determinant
-                  overlap_determinant = det1
-
-                  IF (just_started .AND. penalty_amplitude .lt. 0.0_dp ) THEN 
-                     penalty_occ_vol_prefactor(ispin) = - (-penalty_amplitude)*localization_obj_function_ispin
-                  ENDIF
-                  penalty_func_new = penalty_func_new + penalty_occ_vol_prefactor(ispin) * LOG(det1)
+               CALL compute_obj_nlmos( &
+                  !obj_function_ispin=obj_function_ispin, &
+                  localization_obj_function_ispin=localization_obj_function_ispin, &
+                  penalty_func_ispin=penalty_func_ispin, &
+                  overlap_determinant=overlap_determinant, &
+                  m_sigma=m_sigma(ispin), &
+                  nocc=nocc(ispin), &
+                  m_B0=m_B0(:,:,ispin), &
+                  m_theta_normalized=m_theta_normalized(ispin), &
+                  template_matrix_mo=matrix_mo_out(ispin), &
+                  weights=weights, &
+                  m_S0=m_S0(ispin), &
+                  just_started=just_started, &
+                  penalty_vol_prefactor=penalty_vol_prefactor(ispin), &
+                  penalty_amplitude=penalty_amplitude, &
+                  eps_filter=eps_filter)
 
-               ENDIF
+               localization_obj_function = localization_obj_function + localization_obj_function_ispin
+               penalty_func_new = penalty_func_new + penalty_func_ispin
 
             ENDDO ! ispin
             objf_new = penalty_func_new + localization_obj_function
-            !!!! -- end objective function
+
+!            localization_obj_function = 0.0_dp
```

Comment:
Remove this comment block. Test your refactoring thoroughly before submitting the patch. If it works, no need to keep these comments.